lib/dom/xml_entities.nit

```diff
@@ -0,0 +1,586 @@
+
+	redef fun to_s do
+		var s = "<{tag_name}"
+		if not attributes.is_empty then
+			s += " "
+			s += attributes.join(" ")
+		end
+		s += ">"
+		for i in children do s += i.to_s
+		if matching != null then s += matching.to_s
+		return s
+	end
+end
+
+# Any prolog style-Tag (starting with <?xml)
+class XMLPrologTag
+	super XMLAttrTag
+
+	redef fun to_s do return """<?{{{tag_name}}} {{{attributes.join(" ")}}}?>"""
+end
+
+class XMLProcessingInstructionTag
+	super XMLTag
+
+	var content: String
+
+	redef fun to_s do return "<?{tag_name} {content}?>"
+end
+
+# An end Tag (starting with </)
+class XMLEndTag
+	super XMLTag
+
+	# Optional matching tag, must be matched for the document to be well-formed
+	var matching: nullable XMLStartTag
+
+	redef fun to_s do return "</{tag_name}>"
+end
+
+# An XML comment tag
+class XMLCommentTag
+	super XMLTag
+
+	redef fun to_s do return "<!--{tag_name}-->"
+end
+
+# A DOCTYPE Tag
+class XMLDoctypeTag
+	super XMLTag
+
+	var content: String
+
+	redef fun to_s do return "<!DOCTYPE {content}>"
+end
+
+# A Special Tag (starts with !)
+#
+# TODO: Support the remaining ! tags
+class XMLSpecialTag
+	super XMLTag
+
+	redef fun to_s do return "<!{tag_name}>"
+end
+
+# Attributes are contained in tags, they provide meta-information on a tag
+abstract class XMLAttribute
+	super XMLEntity
+
+	# Name of the attribute
+	var name: String
+end
+
+# An attribute with a String value
+class XMLStringAttr
+	super XMLAttribute
+
+	# Value of the attribute without the double quotes
+	var value: String
+
+	# Type of delimiter (can be either " or ')
+	var delimiter: Char
+
+	redef fun to_s do return "{name}={delimiter}{value}{delimiter}"
+end
+
+# Badly formed XML attribute
+class BadXMLAttribute
+	super XMLAttribute
+
+	redef fun to_s do return name
+end
+
+# Internal use only, shows the end of an attribute block
+private class XMLAttributeEnd
+	super XMLAttribute
+end
+
+# Any XML Error that happens when parsing
+class XMLError
+	super XMLEntity
+
+	# Error message reported by the parser
+	var message: String
+
+	redef fun to_s do return "XML Error: {message} at {location.to_s}"
+end
+
+# Provides XML parsing facilities
+class XMLProcessor
+
+	# Source XML document to parse
+	private var src: String
+
+	# Current position in `src`
+	private var pos = 0
+
+	# Position at which current line started
+	private var line_start = 0
+
+	# Current line in `src`
+	private var line = 1
+
+	# Offset in the current line
+	private fun line_offset: Int do return pos - line_start + 1
+
+	# Advances in `src` until a non-whitespace character is encountered
+	private fun ignore_whitespaces do
+		var srclen = src.length
+		if pos >= srclen then return
+		var c = src[pos]
+		while c.is_whitespace do
+			pos += 1
+			if pos >= srclen then break
+			if c == '\n' then
+				line += 1
+				line_start = pos
+			end
+			c = src[pos]
+		end
+	end
+
+	# Reads characters until pattern `s` is found
+	private fun ignore_until(s: String): Int do
```

Comment:
I thought about it, but there would be some issues with the migration, namely, I want to keep a trace of the line I'm at while parsing, and I can't see a proper way to do this without a second pass, which would render the `search_from` irrelevant from a performance standpoint.
Besides, the search_from yields multiple matches within the whole string, that requires a `substring_from` and I'd only take the first match and ditch the rest, so I don't think it would be much relevant in this case